File: tobas_core/tobas_rqt_tools/src/tobas_rqt_tools/widgets/list_widget.py
from __future__ import (
    annotations,
)  # 自クラスをメソッドの引数としてアノテーションするために必要
from overrides import override
from typing import Union, Iterator
from PyQt5.QtCore import Qt, pyqtSignal
from PyQt5.QtGui import QDropEvent
from PyQt5.QtWidgets import QListWidget, QListWidgetItem


class ListWidget(QListWidget):
    """
    ===== QListWidgetItemとの違い =====
    - イテレータを定義
    - ドラッグアンドドロップでシグナル発行
    - 追加メソッド
    """

    item_moved = pyqtSignal(QListWidgetItem)

    def __iter__(self) -> Iterator[QListWidgetItem]:
        for row in range(self.count()):
            yield self.item(row)

    @override
    def dropEvent(self, event: QDropEvent) -> None:
        super().dropEvent(event)
        self.item_moved.emit(self.selectedItems()[0])

    def remove(self, item: QListWidgetItem) -> None:
        row = self.row(item)
        self.takeItem(row)

    def selected_item(self) -> Union[QListWidgetItem, None]:
        """選択中のアイテムのうち，最も上のものを返す．存在しない場合はNoneを返す．"""
        selected_items = self.selectedItems()
        return selected_items[0] if len(selected_items) > 0 else None


class ListWidgetItem(QListWidgetItem):
    """
    ===== QListWidgetItemとの違い =====
    - UserRoleを基準にソート
    """

    @override
    def __lt__(self, other: ListWidgetItem) -> bool:
        return self.data(Qt.UserRole) < other.data(Qt.UserRole)

File: tobas_gui/tobas_qt_tools/src/widgets/list_widget.cpp
#include "tobas_qt_tools/widgets/list_widget.hpp"

#include <QDebug>
#include <QDropEvent>
#include <QVariant>

namespace qt
{
bool ListWidget::contains(const QString& text) const
{
  const auto items = findItems(text, Qt::MatchExactly);
  return !items.empty();
}

void ListWidget::remove(QListWidgetItem* item)
{
  takeItem(row(item));
}

void ListWidget::setCurrentText(const QString& text)
{
  const auto items = findItems(text, Qt::MatchExactly);
  if (items.empty()) {
    qWarning() << text << "not found.";
    return;
  }

  const auto& item = items.first();
  setCurrentItem(item);
}

void ListWidget::deselect()
{
  const QSignalBlocker block(this);  // 負の行番号でシグナルが発行されるのを防ぐ
  setCurrentRow(-1);
}

void ListWidget::shrinkToContents()
{
  const auto rows = count();
  const auto row_height = sizeHintForRow(0);  // 行の高さ
  const auto frame = 2 * frameWidth();
  const auto margin = contentsMargins().top() + contentsMargins().bottom();

  setFixedHeight(rows * row_height + frame + margin);
  setVerticalScrollBarPolicy(Qt::ScrollBarAsNeeded);
  setSizeAdjustPolicy(QAbstractScrollArea::AdjustToContents);
}

void ListWidget::dropEvent(QDropEvent* event)
{
  QListWidget::dropEvent(event);
  Q_EMIT itemMoved(selectedItems().first());
}

bool ListWidgetItem::operator<(const QListWidgetItem& rhs) const
{
  const auto ldata = this->data(Qt::UserRole);
  const auto rdata = rhs.data(Qt::UserRole);
  return ldata.toString() < rdata.toString();
}
}  // namespace qt
